src/resolvers/exotics/hosted-git-resolver.js

```diff
@@ -201,7 +201,7 @@ export default class HostedGitResolver extends ExoticResolver {
     // If we can access the files over HTTP then we should as it's MUCH faster than git
     // archive and tarball unarchiving. The HTTP API is only available for public repos
     // though.
-    if (await this.hasHTTPCapability(httpUrl)) {
+    if (this.request.getLocked('tarball') || await this.hasHTTPCapability(httpUrl)) {
```

Comment:
It works, but I think it would make more sense semantically speaking to just move [those lines](https://github.com/yarnpkg/yarn/blob/master/src/resolvers/exotics/hosted-git-resolver.js#L143-L146) at the beginning of the `resolve()` function instead.